fabric-networking-api-v1/src/main/java/net/fabricmc/fabric/api/networking/v1/PacketType.java

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright (c) 2016, 2017, 2018, 2019 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.api.networking.v1;
+
+import java.util.function.Function;
+
+import net.minecraft.network.PacketByteBuf;
+import net.minecraft.util.Identifier;
+
+/**
+ * A type of packet. An instance of this should be created per a {@link FabricPacket} implementation.
+ * This holds the channel ID used for the packet.
+ *
+ * <p>Example of creating a packet type:
+ * <pre>{@code
+ * public static final PacketType<BoomPacket> TYPE = PacketType.create(new Identifier("example:boom"), BoomPacket::new);
+ * }</pre>
+ *
+ * @param <T> the type of the packet
+ */
+public final class PacketType<T extends FabricPacket> {
```

Comment:
I want to move the packet write method to a function in the packet type as well, so we can use third-party java classes as packet types.

Reply:
That doesn't really make sense to me. And it's pretty easy to make a record to wrap a java object as a fabric packet.

Reply:
Yeah, the vanilla-like system is more intuitive. (See also BlockEntity, Entity etc, where subclass does serialization)